File: src/boxing_coach/domain/pose.py
"""Pose data model — the output of pose estimation, the input to analysis.

This is deliberately estimator-agnostic. A `PoseSequence` is just keypoints
over time; whether they came from MediaPipe, MMPose, or a synthetic test
fixture is irrelevant to everything downstream.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Iterable, Iterator

import numpy as np

from .landmarks import Landmark


@dataclass(frozen=True, slots=True)
class Keypoint:
    """A single landmark position in image-normalised coordinates."""

    x: float
    y: float
    z: float = 0.0
    visibility: float = 1.0

    @property
    def xy(self) -> np.ndarray:
        return np.array([self.x, self.y], dtype=np.float64)

    @property
    def xyz(self) -> np.ndarray:
        return np.array([self.x, self.y, self.z], dtype=np.float64)


@dataclass(frozen=True, slots=True)
class PoseFrame:
    """All landmarks detected in a single video frame, with a timestamp."""

    index: int
    timestamp_ms: float
    keypoints: dict[Landmark, Keypoint]

    def get(self, landmark: Landmark) -> Keypoint | None:
        return self.keypoints.get(landmark)

    def require(self, landmark: Landmark) -> Keypoint:
        kp = self.keypoints.get(landmark)
        if kp is None:
            raise KeyError(f"frame {self.index} missing landmark {landmark.name}")
        return kp

    def has(self, *landmarks: Landmark, min_visibility: float = 0.0) -> bool:
        return all(
            (kp := self.keypoints.get(lm)) is not None and kp.visibility >= min_visibility
            for lm in landmarks
        )


@dataclass(frozen=True, slots=True)
class PoseSequence:
    """An ordered sequence of pose frames plus capture metadata."""

    frames: list[PoseFrame]
    fps: float
    source: str = "unknown"
    meta: dict[str, object] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if self.fps <= 0:
            raise ValueError("fps must be positive")

    def __len__(self) -> int:
        return len(self.frames)

    def __iter__(self) -> Iterator[PoseFrame]:
        return iter(self.frames)

    @property
    def duration_ms(self) -> float:
        if not self.frames:
            return 0.0
        return self.frames[-1].timestamp_ms - self.frames[0].timestamp_ms

    def trajectory(self, landmark: Landmark) -> np.ndarray:
        """(N, 3) array of xyz for `landmark` across every frame.

        Frames missing the landmark contribute NaN rows so callers can decide
        how to handle gaps rather than silently shifting indices.
        """
        rows: list[np.ndarray] = []
        for frame in self.frames:
            kp = frame.get(landmark)
            rows.append(kp.xyz if kp is not None else np.full(3, np.nan))
        return np.vstack(rows) if rows else np.empty((0, 3))

    def timestamps_ms(self) -> np.ndarray:
        return np.array([f.timestamp_ms for f in self.frames], dtype=np.float64)

    @classmethod
    def from_frames(cls, frames: Iterable[PoseFrame], fps: float, source: str = "unknown") -> "PoseSequence":
        return cls(frames=list(frames), fps=fps, source=source)

File: app/lib/analysis/pose.dart
import 'landmarks.dart';

/// Pose data model — the Dart mirror of `src/boxing_coach/domain/pose.py`.
///
/// Estimator-agnostic: a [PoseSequence] is just keypoints over time, whether
/// they came from MediaPipe on the phone or a golden fixture in a test.

/// A single landmark position in image-normalised coordinates.
class Keypoint {
  const Keypoint(this.x, this.y, {this.z = 0.0, this.visibility = 1.0});

  final double x;
  final double y;
  final double z;
  final double visibility;

  /// [x, y] — the 2D position every v0.5 rule works in.
  List<double> get xy => <double>[x, y];
}

/// All landmarks detected in a single video frame, with a timestamp.
class PoseFrame {
  const PoseFrame({
    required this.index,
    required this.timestampMs,
    required this.keypoints,
  });

  final int index;
  final double timestampMs;
  final Map<Landmark, Keypoint> keypoints;

  Keypoint? get(Landmark landmark) => keypoints[landmark];

  bool has(List<Landmark> landmarks, {double minVisibility = 0.0}) =>
      landmarks.every((lm) {
        final kp = keypoints[lm];
        return kp != null && kp.visibility >= minVisibility;
      });
}

/// An ordered sequence of pose frames plus capture metadata.
class PoseSequence {
  PoseSequence({
    required this.frames,
    required this.fps,
    this.source = 'unknown',
    this.meta = const <String, Object?>{},
  }) {
    if (fps <= 0) throw ArgumentError('fps must be positive');
  }

  final List<PoseFrame> frames;
  final double fps;
  final String source;
  final Map<String, Object?> meta;

  int get length => frames.length;

  double get durationMs => frames.isEmpty
      ? 0.0
      : frames.last.timestampMs - frames.first.timestampMs;

  /// The frame whose timestamp is nearest [ms] — for syncing a skeleton overlay
  /// to the video scrub position. Null only for an empty sequence.
  PoseFrame? frameAtTimestamp(double ms) {
    if (frames.isEmpty) return null;
    var best = frames.first;
    var bestDelta = (best.timestampMs - ms).abs();
    for (final frame in frames) {
      final delta = (frame.timestampMs - ms).abs();
      if (delta < bestDelta) {
        bestDelta = delta;
        best = frame;
      }
    }
    return best;
  }

  /// Encodes to the wire format shared with the Python reference
  /// (`golden_fixtures.sequence_to_json`): landmark index as key, coordinates
  /// rounded to 4 decimals. Used to persist a round's poses alongside its clip.
  Map<String, Object?> toJson() => <String, Object?>{
    'fps': fps,
    'source': source,
    'meta': meta,
    'frames': <Map<String, Object?>>[
      for (final frame in frames)
        <String, Object?>{
          'i': frame.index,
          't': _round4(frame.timestampMs),
          'kp': <String, Object?>{
            for (final entry in frame.keypoints.entries)
              '${entry.key.mpIndex}': <double>[
                _round4(entry.value.x),
                _round4(entry.value.y),
                _round4(entry.value.z),
                _round4(entry.value.visibility),
              ],
          },
        },
    ],
  };

  static double _round4(double v) => (v * 10000).round() / 10000;

  /// Decodes the golden-fixture / calibration-upload wire format. Must match
  /// `golden_fixtures.sequence_to_json` on the Python side exactly.
  factory PoseSequence.fromJson(Map<String, Object?> json) {
    final frameList = (json['frames'] as List<Object?>).map((raw) {
      final frame = raw as Map<String, Object?>;
      final kpMap = frame['kp'] as Map<String, Object?>;
      final keypoints = <Landmark, Keypoint>{};
      kpMap.forEach((indexKey, valuesRaw) {
        final landmark = Landmark.fromIndex(int.parse(indexKey));
        if (landmark == null) return; // a landmark we don't model — skip it
        final values = (valuesRaw as List<Object?>)
            .map((v) => (v as num).toDouble())
            .toList();
        keypoints[landmark] = Keypoint(
          values[0],
          values[1],
          z: values.length > 2 ? values[2] : 0.0,
          visibility: values.length > 3 ? values[3] : 1.0,
        );
      });
      return PoseFrame(
        index: (frame['i'] as num).toInt(),
        timestampMs: (frame['t'] as num).toDouble(),
        keypoints: keypoints,
      );
    }).toList();

    return PoseSequence(
      frames: frameList,
      fps: (json['fps'] as num).toDouble(),
      source: json['source'] as String? ?? 'unknown',
      meta: (json['meta'] as Map<String, Object?>?) ?? const <String, Object?>{},
    );
  }
}
